src/basic/MAT_MAT_SHARED-Cuda.cpp

```diff
@@ -0,0 +1,228 @@
+                                       BODY const &body) {
+  Index_type tx = st + threadIdx.x;
+  if (tx < end)
+    body(tx);
+}
+
+template <typename BODY>
+inline __device__ void thread_y_direct(const Index_type st, const Index_type end,
+                                       BODY const &body) {
+  Index_type ty = st + threadIdx.y;
+  if (ty < end)
+    body(ty);
+}
+
+#define MAT_MAT_SHARED_DATA_SETUP_CUDA                                         \
+  const Index_type NN = getRunSize()*getRunSize();                             \
+  allocAndInitCudaDeviceData(A, m_A, NN);                                      \
+  allocAndInitCudaDeviceData(B, m_B, NN);                                      \
+  allocAndInitCudaDeviceData(C, m_C, NN);
+
+#define MAT_MAT_SHARED_DATA_TEARDOWN_CUDA                                      \
+  getCudaDeviceData(m_A, A, NN);                                               \
+  getCudaDeviceData(m_B, B, NN);                                               \
+  getCudaDeviceData(m_C, C, NN);                                               \
+  deallocCudaDeviceData(A);                                                    \
+  deallocCudaDeviceData(B);                                                    \
+  deallocCudaDeviceData(C);
+
+__global__ void mat_mat_shared(Index_type N, Real_ptr C, Real_ptr A, Real_ptr B) {
+
+  Index_type tx = threadIdx.x;
+  Index_type ty = threadIdx.y;
+  Index_type bx = blockIdx.x;
+  Index_type by = blockIdx.y;
+
+  MAT_MAT_SHARED_BODY_0
+
+  MAT_MAT_SHARED_BODY_1
+
+  for (Index_type k = 0; k < (TL_SZ + N - 1) / TL_SZ; k++) {
+
+    MAT_MAT_SHARED_BODY_2
+
+    __syncthreads();
+
+    MAT_MAT_SHARED_BODY_3
+
+    __syncthreads();
+  }
+
+  MAT_MAT_SHARED_BODY_4
+}
+
+void MAT_MAT_SHARED::runCudaVariant(VariantID vid) {
+
+  const Index_type run_reps = getRunReps();
+  const Index_type N = getRunSize();
+
+  dim3 blockdim(TL_SZ, TL_SZ);
+  dim3 griddim(RAJA_DIVIDE_CEILING_INT(N, blockdim.x),
+               RAJA_DIVIDE_CEILING_INT(N, blockdim.y));
+
+  const Index_type Nx = griddim.x;
+  const Index_type Ny = griddim.y;
+
+  MAT_MAT_SHARED_DATA_SETUP;
+
+  if (vid == Base_CUDA) {
+
+    MAT_MAT_SHARED_DATA_SETUP_CUDA;
+
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+
+      mat_mat_shared<<<griddim, blockdim>>>(N, A, B, C);
+    }
+    stopTimer();
+
+    MAT_MAT_SHARED_DATA_TEARDOWN_CUDA;
+
+  } else if (vid == Lambda_CUDA) {
+
+    MAT_MAT_SHARED_DATA_SETUP_CUDA;
+
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+
+      lambda_cuda<<<griddim, blockdim>>>([=] __device__() {
+        block_y_direct(0, Ny, [&](Index_type by) {
+          block_x_direct(0, Nx, [&](Index_type bx) {
+            MAT_MAT_SHARED_BODY_0
+
+            thread_y_direct(0, TL_SZ, [&](Index_type ty) {
+              thread_x_direct(0, TL_SZ, [&](Index_type tx) { MAT_MAT_SHARED_BODY_1 });
+            });
```

Comment:
I'm not sure that using functions to do the mapping here quite fits with the way things are done in the rest of the perf suite. Shouldn't this be the same as the direct cuda version with just the addition of lambdas?
```suggestion
            Index_type ty = threadIdx.y;
            Index_type tx = threadIdx.x;
            auto outer_body_1 = [&](Index_type ty) {
               auto inner_body_1 = [&](Index_type tx) { MAT_MAT_SHARED_BODY_1 };
               if (tx < TL_SZ) inner_body_1(tx);
            };
            if (ty < TL_SZ) outer_body_1(ty);
```

Reply:
I was thinking it would be easier to read, but that's a good point. I'll switch to this style